File: mtgSearch/src/main/java/com/dbottillo/lifecounter/LifeCounterFragment.java
package com.dbottillo.lifecounter;

import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.animation.PropertyValuesHolder;
import android.annotation.TargetApi;
import android.app.Activity;
import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.SharedPreferences;
import android.graphics.Outline;
import android.os.Build;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewOutlineProvider;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.ScrollView;
import android.widget.Toast;

import com.dbottillo.R;
import com.dbottillo.adapters.LifeCounterAdapter;
import com.dbottillo.base.DBFragment;
import com.dbottillo.database.DB40Helper;
import com.dbottillo.helper.DBAsyncTask;
import com.dbottillo.helper.TrackingHelper;
import com.dbottillo.resources.Player;
import com.dbottillo.util.AnimationUtil;

import java.util.ArrayList;
import java.util.Locale;
import java.util.Random;

import fr.castorflex.android.smoothprogressbar.SmoothProgressBar;

public class LifeCounterFragment extends DBFragment implements DBAsyncTask.DBAsyncTaskListener, LifeCounterAdapter.OnLifeCounterListener, View.OnClickListener {

    public static LifeCounterFragment newInstance() {
        return new LifeCounterFragment();
    }

    private ScrollView diceScrollView;
    private LinearLayout diceContainer;
    private ListView lifeListView;
    private ArrayList<Player> players;
    private LifeCounterAdapter lifeCounterAdapter;
    private SmoothProgressBar progressBar;
    private FloatingActionButton newPlayerButton;

    private DB40Helper db40Helper;

    private boolean scrollDownAfterLoad = false;

    private boolean showPoison = false;
    private boolean diceShowed = false;
    private boolean twoHGEnabled = false;

    String[] names = {"Teferi", "Nicol Bolas", "Gerrard", "Ajani", "Jace", "Liliana", "Elspeth", "Tezzeret", "Garruck",
            "Chandra", "Venser", "Doran", "Sorin"};

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_life_counter, container, false);

        setActionBarTitle(getString(R.string.action_life_counter));

        diceScrollView = (ScrollView) rootView.findViewById(R.id.life_counter_dice_scrolliew);
        diceContainer = (LinearLayout) rootView.findViewById(R.id.life_counter_dice_container);
        progressBar = (SmoothProgressBar) rootView.findViewById(R.id.progress);
        lifeListView = (ListView) rootView.findViewById(R.id.life_counter_list);
        showPoison = getSharedPreferences().getBoolean("poison", false);

        View footerView = inflater.inflate(R.layout.fab_button_list_footer, lifeListView, false);
        lifeListView.addFooterView(footerView);

        twoHGEnabled = getSharedPreferences().getBoolean(PREF_TWO_HG_ENABLED, false);

        newPlayerButton = (FloatingActionButton) rootView.findViewById(R.id.new_player);
        newPlayerButton.setOnClickListener(this);

        players = new ArrayList<>();

        lifeCounterAdapter = new LifeCounterAdapter(getActivity(), players, this, showPoison);
        lifeListView.setAdapter(lifeCounterAdapter);

        AnimationUtil.growView(newPlayerButton);

        setHasOptionsMenu(true);

        return rootView;
    }

    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        db40Helper = DB40Helper.getInstance(activity);
    }

    @Override
    public void onStart() {
        super.onStart();

        db40Helper.openDb();
        progressBar.setVisibility(View.VISIBLE);
        loadPlayers();
    }

    @Override
    public void onStop() {
        super.onStop();

        db40Helper.closeDb();
    }

    @Override
    public void onResume() {
        super.onResume();
        setScreenOn(getSharedPreferences().getBoolean(PREF_SCREEN_ON, false));
    }

    private void setScreenOn(boolean screenOn) {
        if (getView() != null) {
            getView().setKeepScreenOn(screenOn);
        }
    }

    private void resetLifeCounter() {
        for (Player player : players) {
            player.setLife(twoHGEnabled ? 30 : 20);
            player.setPoisonCount(twoHGEnabled ? 15 : 10);
            db40Helper.storePlayer(player);
        }
        loadPlayers();
    }


    @Override
    public String getPageTrack() {
        return "/life_counter";
    }

    private void loadPlayers() {
        new DBAsyncTask(getActivity(), this, DBAsyncTask.TASK_PLAYER).execute();
    }

    private void addPlayer() {
        if (players.size() == 10) {
            Toast.makeText(getActivity(), R.string.maximum_player, Toast.LENGTH_SHORT).show();
            return;
        }
        Player player = new Player(getUniqueIdForPlayer(), getUniqueNameForPlayer());
        db40Helper.storePlayer(player);
        scrollDownAfterLoad = true;
        loadPlayers();
    }

    private String getUniqueNameForPlayer() {
        boolean unique = false;
        int pickedNumber = 0;
        while (!unique) {
            Random rand = new Random();
            pickedNumber = rand.nextInt(names.length);
            boolean founded = false;
            for (Player player : players) {
                if (player.getName().toLowerCase(Locale.getDefault()).contains(names[pickedNumber].toLowerCase(Locale.getDefault()))) {
                    founded = true;
                    break;
                }
            }
            if (!founded) unique = true;
        }
        return names[pickedNumber];
    }

    private int getUniqueIdForPlayer() {
        int id = 0;
        for (Player player : players) {
            if (id == player.getId()) {
                id++;
            }
        }
        return id;
    }

    @Override
    public void onTaskFinished(int type, ArrayList<?> objects) {
        progressBar.setVisibility(View.GONE);
        if (objects.size() == 0) {
            // need at least one player
            addPlayer();
        } else {
            players.clear();
            for (Object player : objects) {
                //Log.e("magic", "loaded: "+((Player)player).getId()+" - "+((Player)player).toString());
                players.add((Player) player);
            }
        }
        lifeCounterAdapter.notifyDataSetChanged();
        if (scrollDownAfterLoad) {
            ((ListView) getView().findViewById(R.id.life_counter_list)).setSelection(players.size() - 1);
        }
        scrollDownAfterLoad = false;
    }

    @Override
    public void onTaskEndWithError(int type, String error) {
        progressBar.setVisibility(View.GONE);
        Toast.makeText(getActivity(), error, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onRemovePlayer(int position) {
        TrackingHelper.getInstance(getActivity()).trackEvent(TrackingHelper.UA_CATEGORY_LIFE_COUNTER, "removePlayer");
        db40Helper.removePlayer(players.get(position));
        loadPlayers();
    }

    @Override
    public void onEditPlayer(final int position) {
        AlertDialog.Builder alert = new AlertDialog.Builder(getActivity());

        alert.setTitle(getString(R.string.edit_player));

        final EditText input = new EditText(getActivity());
        input.setText(players.get(position).getName());
        alert.setView(input);

        alert.setPositiveButton(getString(R.string.save), new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int whichButton) {
                String value = input.getText().toString();
                players.get(position).setName(value);
                db40Helper.storePlayer(players.get(position));
                loadPlayers();
                TrackingHelper.getInstance(getActivity()).trackEvent(TrackingHelper.UA_CATEGORY_LIFE_COUNTER, "editPlayer");
            }
        });

        alert.setNegativeButton(getString(R.string.cancel), new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int whichButton) {
                // Canceled.
            }
        });

        alert.show();

    }

    @Override
    public void onLifeCountChange(int position, int value) {
        TrackingHelper.getInstance(getActivity()).trackEvent(TrackingHelper.UA_CATEGORY_LIFE_COUNTER, "lifeCountChanged");
        players.get(position).changeLife(value);
        db40Helper.storePlayer(players.get(position));
        loadPlayers();
    }

    @Override
    public void onPoisonCountChange(int position, int value) {
        TrackingHelper.getInstance(getActivity()).trackEvent(TrackingHelper.UA_CATEGORY_LIFE_COUNTER, "poisonCountChange");
        players.get(position).changePoisonCount(value);
        db40Helper.storePlayer(players.get(position));
        loadPlayers();
    }

    private View.OnClickListener tapOnDice = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            hideDice();
        }
    };

    private void launchDice() {
        if (diceShowed) {
            for (int i = 0; i < players.size(); i++) {
                Player player = players.get(i);
                player.setDiceResult(-1);
            }
        } else {
        /*diceScrollView.setVisibility(View.VISIBLE);
        diceContainer.removeAllViews();
        int heightRow = lifeListView.getChildAt(0).getHeight();
        ArrayList<TextView> playerResults = new ArrayList<TextView>(players.size());*/
            //int[] results = new int[players.size()];
            for (int i = 0; i < players.size(); i++) {
                Player player = players.get(i);
                Random rand = new Random();
                //results[i] = rand.nextInt(20) + 1;
                player.setDiceResult(rand.nextInt(20) + 1);
           /* View dice = LayoutInflater.from(getActivity()).inflate(R.layout.life_counter_dice, null);
            dice.setLayoutParams(new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, heightRow));
            TextView playerName = (TextView) dice.findViewById(R.id.player_name);
            playerName.setText(player.getName());
            playerResults.add((TextView) dice.findViewById(R.id.player_result));
            diceContainer.addView(dice);
            dice.setOnClickListener(tapOnDice);*/
            }
        }
       /* Button closeBtn = new Button(getActivity(), null, R.style.BtnGeneric);
        closeBtn.setBackgroundResource(R.drawable.btn_common);
        closeBtn.setTextColor(getResources().getColor(android.R.color.white));
        LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT);
        params.topMargin = 20;
        params.gravity = Gravity.CENTER_HORIZONTAL;
        closeBtn.setLayoutParams(params);
        closeBtn.setText(getString(R.string.dice_close));
        closeBtn.setPadding(20, 20, 20, 20);
        closeBtn.setOnClickListener(tapOnDice);
        diceContainer.addView(closeBtn);
        for (int i = 0; i < players.size(); i++) {
            playerResults.get(i).setText(results[i] + "");
        }
        diceShowed = true;*/
        diceShowed = !diceShowed;
        lifeCounterAdapter.notifyDataSetChanged();
    }

    private void hideDice() {
        diceContainer.removeAllViews();
        diceScrollView.setVisibility(View.GONE);
        diceShowed = false;
    }

    @Override
    public void onClick(View v) {
        addPlayer();
        TrackingHelper.getInstance(getActivity()).trackEvent(TrackingHelper.UA_CATEGORY_LIFE_COUNTER, "addPlayer");
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.life_counter, menu);

        MenuItem poison = menu.findItem(R.id.action_poison);
        if (showPoison) {
            poison.setChecked(true);
        } else {
            poison.setChecked(false);
        }

        MenuItem screenOn = menu.findItem(R.id.action_screen_on);
        screenOn.setChecked(getSharedPreferences().getBoolean(PREF_SCREEN_ON, false));

        MenuItem twoHg = menu.findItem(R.id.action_two_hg);
        twoHg.setChecked(twoHGEnabled);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int i1 = item.getItemId();
        if (i1 == R.id.action_reset) {
            resetLifeCounter();
            TrackingHelper.getInstance(getActivity()).trackEvent(TrackingHelper.UA_CATEGORY_LIFE_COUNTER, "resetLifeCounter");
            return true;
        }
        if (i1 == R.id.action_dice) {
            launchDice();
            TrackingHelper.getInstance(getActivity()).trackEvent(TrackingHelper.UA_CATEGORY_LIFE_COUNTER, "launchDice");
            return true;
        }
        if (i1 == R.id.action_poison) {
            TrackingHelper.getInstance(getActivity()).trackEvent(TrackingHelper.UA_CATEGORY_LIFE_COUNTER, "poisonSetting");
            getSharedPreferences().edit().putBoolean("poison", !showPoison).apply();
            showPoison = !showPoison;
            getActivity().invalidateOptionsMenu();
            lifeCounterAdapter.setShowPoison(showPoison);
            lifeCounterAdapter.notifyDataSetChanged();
            return true;
        }
        if (i1 == R.id.action_screen_on) {
            boolean screenOn = getSharedPreferences().getBoolean(PREF_SCREEN_ON, false);
            SharedPreferences.Editor editor = getSharedPreferences().edit();
            editor.putBoolean(PREF_SCREEN_ON, !screenOn);
            editor.apply();
            getActivity().invalidateOptionsMenu();
            setScreenOn(!screenOn);
            TrackingHelper.getInstance(getActivity()).trackEvent(TrackingHelper.UA_CATEGORY_LIFE_COUNTER, "screenOn");
        }
        if (i1 == R.id.action_two_hg) {
            twoHGEnabled = !twoHGEnabled;
            SharedPreferences.Editor editor = getSharedPreferences().edit();
            editor.putBoolean(PREF_TWO_HG_ENABLED, twoHGEnabled);
            editor.apply();
            getActivity().invalidateOptionsMenu();
            resetLifeCounter();
            TrackingHelper.getInstance(getActivity()).trackEvent(TrackingHelper.UA_CATEGORY_LIFE_COUNTER, "two_hg");
        }

        return false;
    }

}

File: feature_life_counter/src/main/kotlin/com/dbottillo/mtgsearchfree/lifecounter/LifeCounterFragment.kt
package com.dbottillo.mtgsearchfree.lifecounter

import android.content.Context
import android.os.Bundle
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.dbottillo.mtgsearchfree.exceptions.MTGException
import com.dbottillo.mtgsearchfree.model.Player
import com.dbottillo.mtgsearchfree.storage.CardsPreferences
import com.dbottillo.mtgsearchfree.home.BaseHomeFragment
import com.dbottillo.mtgsearchfree.util.DialogUtil
import com.dbottillo.mtgsearchfree.util.LOG
import com.dbottillo.mtgsearchfree.util.TrackingManager
import dagger.android.support.AndroidSupportInjection
import java.util.Random
import javax.inject.Inject

class LifeCounterFragment : BaseHomeFragment(), LifeCounterView, OnLifeCounterListener {

    private lateinit var lifeCounterList: RecyclerView

    @Inject lateinit var lifeCounterPresenter: LifeCounterPresenter
    @Inject lateinit var cardsPreferences: CardsPreferences
    @Inject lateinit var dialogUtil: DialogUtil

    private lateinit var adapter: LifeCounterAdapter
    private var players: MutableList<Player> = mutableListOf()
    private var diceShowed: Boolean = false

    override fun onAttach(context: Context) {
        AndroidSupportInjection.inject(this)
        super.onAttach(context)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val rootView = inflater.inflate(R.layout.fragment_life_counter, container, false)
        dialogUtil.init(context)
        return rootView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        lifeCounterList = view.findViewById(R.id.life_counter_list)
        view.findViewById<View>(R.id.action_reset).setOnClickListener { reset() }
        view.findViewById<View>(R.id.action_dice).setOnClickListener { launchDice() }
        view.findViewById<View>(R.id.add_new_player).setOnClickListener { addPlayer() }

        lifeCounterList.setHasFixedSize(true)
        lifeCounterList.layoutManager = LinearLayoutManager(view.context)

        setupMenu()

        adapter = LifeCounterAdapter(players, this, cardsPreferences.showPoison())
        lifeCounterList.adapter = adapter

        lifeCounterPresenter.init(this)
        lifeCounterPresenter.loadPlayers()
    }

    override fun onResume() {
        super.onResume()
        setScreenOn(cardsPreferences.screenOn())
    }

    override fun onPause() {
        super.onPause()
        setScreenOn(false)
    }

    override fun getPageTrack(): String {
        return "/life_counter"
    }

    override fun getScrollViewId() = R.id.life_counter_list
    override fun getToolbarId() = R.id.toolbar
    override fun getToolbarTitleId() = R.id.toolbar_title

    override fun getTitle(): String {
        return resources.getString(R.string.action_life_counter)
    }

    override fun showError(message: String) {
        Toast.makeText(activity, message, Toast.LENGTH_SHORT).show()
    }

    override fun playersLoaded(newPlayers: List<Player>) {
        players.clear()
        players.addAll(newPlayers)

        adapter.notifyDataSetChanged()
    }

    override fun showError(exception: MTGException) {
        Toast.makeText(activity, exception.localizedMessage, Toast.LENGTH_SHORT).show()
    }

    override fun onLifeCountChange(player: Player, value: Int) {
        TrackingManager.trackLifeCountChanged()
        player.changeLife(value)
        lifeCounterPresenter.editPlayer(player)
    }

    override fun onPoisonCountChange(player: Player, value: Int) {
        LOG.d()
        TrackingManager.trackPoisonCountChanged()
        player.changePoisonCount(value)
        lifeCounterPresenter.editPlayer(player)
    }

    override fun onEditPlayer(player: Player) {
        LOG.d()
        dialogUtil.showEditPlayer(R.layout.dialog_edit_deck, R.id.edit_text, player) {
            player.name = it
            lifeCounterPresenter.editPlayer(player)
            TrackingManager.trackEditPlayer()
        }
    }

    override fun onRemovePlayer(player: Player) {
        LOG.d()
        lifeCounterPresenter.removePlayer(player)
        TrackingManager.trackRemovePlayer()
    }

    private fun addPlayer() {
        LOG.d()
        if (lifeCounterList.adapter?.itemCount == 10) {
            Toast.makeText(activity, R.string.maximum_player, Toast.LENGTH_SHORT).show()
            return
        }
        lifeCounterPresenter.addPlayer()
        TrackingManager.trackAddPlayer()
    }

    private fun setupMenu() {
        toolbar.inflateMenu(R.menu.life_counter)

        refreshMenu()
    }

    private fun refreshMenu() {
        val poison = toolbar.menu.findItem(R.id.action_poison)
        poison.isChecked = cardsPreferences.showPoison()
        val screenOn = toolbar.menu.findItem(R.id.action_screen_on)
        screenOn.isChecked = cardsPreferences.screenOn()
        val twoHg = toolbar.menu.findItem(R.id.action_two_hg)
        twoHg.isChecked = cardsPreferences.twoHGEnabled()
    }

    override fun onMenuItemClick(item: MenuItem?): Boolean {
        when (item?.itemId) {
            R.id.action_poison -> poisonChanged()
            R.id.action_screen_on -> screenOnChanged()
            R.id.action_two_hg -> twoHGChanged()
            else -> super.onMenuItemClick(item)
        }
        return true
    }

    private fun twoHGChanged() {
        val twoHGEnabled = !cardsPreferences.twoHGEnabled()
        cardsPreferences.setTwoHGEnabled(twoHGEnabled)
        refreshMenu()
        resetLifeCounter()
        TrackingManager.trackHGLifeCounter()
    }

    private fun poisonChanged() {
        TrackingManager.trackChangePoisonSetting()
        val showPoison = !cardsPreferences.showPoison()
        cardsPreferences.showPoison(showPoison)
        refreshMenu()
        adapter.setShowPoison(showPoison)
        adapter.notifyDataSetChanged()
    }

    private fun screenOnChanged() {
        val screenOn = cardsPreferences.screenOn()
        cardsPreferences.setScreenOn(!screenOn)
        refreshMenu()
        setScreenOn(!screenOn)
        TrackingManager.trackScreenOn()
    }

    private fun setScreenOn(screenOn: Boolean) {
        LOG.d()
        if (view != null) {
            view!!.keepScreenOn = screenOn
        }
    }

    private fun resetLifeCounter() {
        LOG.d()
        val twoHGEnabled = cardsPreferences.twoHGEnabled()
        for (player in players) {
            player.life = if (twoHGEnabled) 30 else 20
            player.poisonCount = if (twoHGEnabled) 15 else 10
        }
        lifeCounterPresenter.editPlayers(players)
    }

    fun reset() {
        resetLifeCounter()
        TrackingManager.trackResetLifeCounter()
    }

    private fun launchDice() {
        if (diceShowed) {
            players.forEach { it.diceResult = -1 }
        } else {
            players.forEach { it.diceResult = Random().nextInt(20) + 1 }
        }
        diceShowed = !diceShowed
        adapter.notifyDataSetChanged()
        TrackingManager.trackLunchDice()
    }

    override fun onDestroy() {
        super.onDestroy()
        lifeCounterPresenter.onDestroy()
    }
}
